### PR TITLE
Test launch.m

### PR DESCRIPTION
**Pull request by j-emberton**
PR for testing and refactor of launch.m

progress to date:
- added regression test framework for launch.m testing against reference outputs for chosen test cases
- added unit tests which test against exitcode output for a variety of input scenarios
- reworked some global fixtures in conftest to better suit test requirements
- added pyproject.toml to setup desired pytest behaviour
- tests run successfully

Decisions to be made:
- do we want to extend range of unit tests by capturing stdout similar to regression test? Would allow greater test coverage. Do we want to extend print statements in launch.m to cover a greater range of scenarios?
- Once we agree tests, I'll start the refactor

NB:
- At the moment this branch is based on test_infrastructure branch rather than the ustar_cp_refactor_master branch. Once we merge the test_infrastructure PR into master, I'll rebase this branch to tidy things up. There's a lot of duplication in the commits and the diff just now which is a bit untidy.

### CODE REVIEW

**Comment by j-emberton** [comment]
Added coverage check. Doesn't actually work on the matlab code (only covers python and its tests) but will be useful once we start moving into Python

**Comment by dorchard** [comment]
Thanks for adding the extra documentation - this is looking good!

**Comment by dorchard** [comment]
> Thanks for adding the extra documentation - this is looking good!

Sorry I added this to the wrong PR. Will look at this in depth now.

**Comment by dorchard** [comment]
Retargetting this PR to just be about the tests for launch.m rather than any refactoring yet. Now needs rebasing off `ustar_cp_refactor_main`

**Comment by j-emberton** [comment]
Rebased successfully

**Comment by j-emberton** [comment]
Going to aim to drop aim to refactor in this PR while we sort out integrating other test cases with ifs.

These tests make a decent baseline we can build on.

**Comment by dorchard** [comment]
I can confirm that everything is running fine for me now with the tests. What's not clear is whether this PR plans to include all the end-to-end data tests or not over `launch.m`? Or will we add that as a separate PR?

**Comment by j-emberton** [comment]
> I can confirm that everything is running fine for me now with the tests. What's not clear is whether this PR plans to include all the end-to-end data tests or not over `launch.m`? Or will we add that as a separate PR?

Good to hear!

What do you mean by 'end-to-end data tests'? (Sorry, I might be being slow!).

The regression test has a single 'end-to-end' test but I'd agree this is incomplete as it doesn't give us good code coverage. I'd propose that we select out several cases to use as our 'reference' cases in a separate PR (partly base we have 400MB of test data and we maybe don't need all of it, just a small subset. This would also solve the issue of where do we store 400MB of data as git LFS isn't an option).

We should get all the end to end tests in place before thinking about merging any refactored code.

We might want to think about whether we want to do an initial refactor aimed at improving testability (i.e. get it to return more than 1 or 0) and a 2nd refactor to modularise it.

**Comment by j-emberton** [comment]
@isaacaka @tztsai 

I hope to merge this PR in the next team meeting (12th Sept). Could you review this PR and make any suggestions in the next day or two so that we have time to implement and review any changes prior to the 12th.

File: tests/conftest.py [hunk 1]
```diff
@@ -46,37 +47,163 @@ def matlab_engine(refactored = False):
...
+
+    # Cleanup just in case
+    for folder in [input_folder, reference_output_folder, output_folder]:
+        if folder.exists():
+            shutil.rmtree(folder)
+
+    # Create the output directory (starts empty)
+    output_folder.mkdir()
+
+    # Pattern to match directories starting with 'US_ARc'
```
Comment (dorchard) [inline tests/conftest.py]:
```suggestion
    # Pattern to match directories starting with the `testcase` name
```

File: tests/regression_tests/test_ustar_cp/test_ustar_cp.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+The tests are parameterized to allow easy addition of new test cases and to automate the 
+comparison of MATLAB's output with expected results. The focus is on validating the standard 
+output and the exit code returned by the MATLAB function.
+"""
+
+import pytest
+import io
+
+test_cases = [
+    ("US_ARc", [1]),
```
Comment (dorchard) [inline tests/regression_tests/test_ustar_cp/test_ustar_cp.py]: Are we missing the rest of the values here? `test_ustar_cp` expects 6 arguments but this a 2-tuple. Am I missing something?

File: tests/unit_tests/test_ustar_cp/test_launch.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        The test asserts that the MATLAB function returns an exit code of 0, indicating 
+        that the function identified the missing file scenario.
+    """
+    _, _, empty_output = setup_folders
+
+    # Run the MATLAB function
+    exitcode = matlab_engine.launch(empty_output, empty_output)
+
+    # Check that the exitcode indicates an error
+    assert exitcode == 0, "Expected zero exitcode for missing file."
```
Comment (dorchard) [inline tests/unit_tests/test_ustar_cp/test_launch.py:82]: It surprises me that this isn't a non-zero value. The comment here is then also a bit confusing since 0 means no failure.

File: tests/regression_tests/test_ustar_cp/test_ustar_cp.py (hunk 2 above)
Comment (j-emberton) [reply tests/regression_tests/test_ustar_cp/test_ustar_cp.py]: Yeah, test_ustar_cp takes 6 args. Args 1 and 2 are the test case and output (supplied by the prior parameterisation), the remainder are pytest fixtures which are used to process the test case data. These don't need to be specified as inputs in the associated `parametrize` declaration. They are available by virtue of being declared in the conftest.py or in the same module.

Comment (dorchard) [reply tests/regression_tests/test_ustar_cp/test_ustar_cp.py]: Ah yes sorry. I get it. Thanks.

File: tests/unit_tests/test_ustar_cp/test_launch.py (hunk 3 above)
Comment (j-emberton) [reply tests/unit_tests/test_ustar_cp/test_launch.py:82]: The exit code is zero for a missing file input because the code doesn't update the exit code if the folder is missing.

`% check input path
if 0 == exist('input_folder')
    input_folder = [pwd '/'];
elseif length(input_folder) < 2
    if input_folder(2) ~= ':'
        input_folder = [pwd '/' input_folder];
    end
end
if input_folder(length(input_folder)) ~= '\' && input_folder(length(input_folder)) ~= '/'
    input_folder = [input_folder '/'];
end`

This is the section of code from launch.m that is being tested. At the moment, if an input folder is not passed, it looks in the current working directory and uses that.

So not passing an input folder is technically not an error state hence the code doesn't update the exit code to 1.

Later in the code, the number of files in the working dir matching a pattern is evaluated. If the number is zero i.e. no files are found, the loop simply isn't runand the function exits with code zero. 

It does however print a line that  states 0 files found. Maybe we should do a test for that?

Comment (dorchard) [reply tests/unit_tests/test_ustar_cp/test_launch.py:82]: Okay thanks for that explanation. I think that the comment on line 72 could be less confusing then. And yes, I think we could test the output (e.g., like in this test: https://github.com/Cambridge-ICCS/ONEFlux/pull/25/files#diff-770a227d2135b6206c5cbdd0a928848a2a3b07282e242ffc2ce82296e34b1a03R176-R183)

Comment (j-emberton) [reply tests/unit_tests/test_ustar_cp/test_launch.py:82]: OK, fixed the docstring and comments and added extra assertion to test for std output under the empty folder condition

Comment (dorchard) [reply tests/unit_tests/test_ustar_cp/test_launch.py:82]: great thanks!

**Review by dorchard — APPROVED** [review]

File: tests/conftest.py [hunk 4]
```diff
@@ -46,37 +47,163 @@ def matlab_engine(refactored = False):
...
+
+    Args:
+        text1 (str): The first block of text.
+        text2 (str): The second block of text.
 
-    return str(input_folder), str(output_folder)
+    Returns:
+        bool: True if the stripped text blocks are identical, False otherwise.
+    """
+    return text1.strip() == text2.strip()
```
Comment (isaacaka) [inline tests/conftest.py]: May want to handle the case for removing new lines for if/when we compare longer output: e.g. adding `.replace('\n', '')`

**Comment by isaacaka** [comment]
This looks good to me overall. Nice work!

**Comment by isaacaka** [comment]
This looks good to me. Nice work!

File: tests/unit_tests/test_ustar_cp/test_launch.py (hunk 3 above)
Comment (casterkay) [reply tests/unit_tests/test_ustar_cp/test_launch.py:82]: Maybe you can add a `StringIO` here to capture the stdout for testing the number of files detected.

File: tests/unit_tests/test_ustar_cp/test_launch.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    Asserts:
+        The test asserts that the MATLAB function returns a non-zero exit code, indicating 
+        that it identified and handled the invalid data correctly.
+    """
+    input_folder, output_folder = setup_test_environment
+    eng = matlab_engine
+
+    # Overwrite the input file with invalid data
+    invalid_data = """invalid content"""
```
Comment (casterkay) [inline tests/unit_tests/test_ustar_cp/test_launch.py:109]: Is it necessary to use multiline string here?

**Review by casterkay — APPROVED** [review]

**Review by isaacaka — APPROVED** [review]

File: tests/conftest.py (hunk 4 above)
Comment (isaacaka) [inline tests/conftest.py]:
```suggestion
    return text1.replace('\n', '').strip() == text2.replace('\n', '').strip()
```

**Review by isaacaka — APPROVED** [review]
